### PR TITLE
DCOS-8808 & DCOS-8904: rethink the struct approach in the service form modal

### PR DESCRIPTION
**Pull request by Poltergeist**
This PR introduces a `getJSON` method to the Item struct which will return a JSON of the `_itemData` this enables us to

- Make unknown fields which have been added in JSON mode persistent after switching back to the form.
- Keep fields in edit mode which are not present in the form, for example `requirePorts`, `updateStrategy`, `residency` to name a few. Those are also exposed to the JSON mode. 
- Remove fields which have been set in an earlier version for example `user` right now it is only possible to rename the user but not to remove it from the form. because if the value is `null` the field was removed.

In general the JSON is now a lot more verbose but it is necessary.

image

image

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @mlunoe

File: src/js/constants/blacklistServiceConfig.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+let BlackListServiceConfig = [
```
Comment (orlandohohmeier) [inline src/js/constants/blacklistServiceConfig.js]: Let's rename this file and provide the black list as a variable so that it is consistent with the other constant files.  How about `ServiceConfig.BLACK_LIST`?

File: src/js/structs/ServiceTree.js [hunk 2]
```diff
@@ -48,6 +48,9 @@ module.exports = class ServiceTree extends Tree {
 
       // Check the DCOS_PACKAGE_FRAMEWORK_NAME label to determine if the item
       // should be converted to an Application or Framework instance.
+      if (item instanceof Framework || item instanceof Application) {
```
Comment (orlandohohmeier) [inline src/js/structs/ServiceTree.js]: Let's add this checks to the condition in line 37 https://github.com/dcos/dcos-ui/pull/903/files#diff-0eb72bc636980e431e5fe14b2a54616dR37

File: src/js/utils/ServiceUtil.js [hunk 3]
```diff
@@ -2,6 +2,25 @@ import {Hooks} from 'PluginSDK';
 import Service from '../structs/Service';
 import VolumeConstants from '../constants/VolumeConstants';
 
+function isEmpty(data) {
```
Comment (orlandohohmeier) [inline src/js/utils/ServiceUtil.js]: Can we move this to the validator util?

File: src/js/structs/Item.js [hunk 4]
```diff
@@ -7,6 +7,9 @@ module.exports = class Item {
     this._itemData = item;
   }
 
+  toJSON() {
+    return JSON.stringify(this._itemData);
```
Comment (mlunoe) [inline src/js/structs/Item.js]: Let's do `this.get()` to not access the internal property directly and use the API in stead.

File: src/js/structs/ServiceTree.js (hunk 2 above)
Comment (mlunoe) [inline src/js/structs/ServiceTree.js]: Let's check for `Service` here

File: src/js/utils/ServiceUtil.js [hunk 5]
```diff
@@ -2,6 +2,25 @@ import {Hooks} from 'PluginSDK';
...
+  if (typeof(data) == 'number' || typeof(data) == 'boolean') {
```
Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: We use `typeof data` elsewhere, let's stick to that, or change this everywhere

File: src/js/utils/ServiceUtil.js [hunk 6]
```diff
@@ -2,6 +2,25 @@ import {Hooks} from 'PluginSDK';
...
+    return false;
+  }
+  if (typeof(data) == 'undefined' || data === null) {
+    return true;
+  }
+  if (typeof(data.length) != 'undefined') {
+    return data.length === 0;
+  }
+  var count = 0;
```
Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: `let`

File: src/js/utils/ServiceUtil.js [hunk 7]
```diff
@@ -2,6 +2,25 @@ import {Hooks} from 'PluginSDK';
...
+  for (var i in data) {
```
Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: let's use a `reduce` for this

File: src/js/utils/ServiceUtil.js (hunk 3 above)
Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: Let's get some tests for this function

Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: Also, we should add this to `Util`

Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: Haha, just saw your note now @orlandohohmeier 😆

File: src/js/utils/ServiceUtil.js [hunk 8]
```diff
@@ -174,7 +192,7 @@ const ServiceUtil = {
           if (externalVolumes.length > 0) {
             volumesList = volumesList.concat(externalVolumes);
           }
-          if (!isEdit) {
+          if (!isEdit && volumesList.length > 0) {
```
Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: `&& volumesList.length` should be enough

File: src/js/constants/BlackListServiceConfig.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+let BlackListServiceConfig = [
```
Comment (orlandohohmeier) [inline src/js/constants/BlackListServiceConfig.js]: Let's rename this file and provide the black list as a variable so that it is consistent with the other constant files. How about ServiceConfig.BLACK_LIST?

File: src/js/utils/ServiceUtil.js [hunk 10]
```diff
@@ -364,6 +382,15 @@ const ServiceUtil = {
       }
     }
 
+    definition = Object.keys(definition).reduce(function (memo, key) {
+      if (!isEmpty(definition[key])) {
+        memo[key] = definition[key];
+      } else {
+        memo[key] = null;
+      }
+      return memo;
```
Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: #nit newline before `return`

File: src/js/components/modals/ServiceFormModal.js [hunk 11]
```diff
@@ -103,6 +104,15 @@ const responseAttributePathToFieldIdMap = {
   '/': 'general'
 };
 
+var cleanJSONdefinition = function (jsonDefinition) {
+  return Object.keys(jsonDefinition).filter(function (key) {
+    return !BlackListServiceConfig.includes(key);
+  }).reduce(function (memo, key) {
+    memo[key] = jsonDefinition[key];
+    return memo;
```
Comment (mlunoe) [inline src/js/components/modals/ServiceFormModal.js]: #nit newline before `return`

File: src/js/constants/BlackListServiceConfig.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  'version',
+  'versionInfo',
+  'deployments',
+  'tasks',
+  'taskStats',
+  'tasksHealthy',
+  'tasksRunning',
+  'tasksStaged',
+  'tasksUnhealthy',
+  '_itemData'
```
Comment (mlunoe) [inline src/js/constants/BlackListServiceConfig.js]: It this necessary? Isn't this an effect of the bug you discovered here: https://github.com/dcos/dcos-ui/pull/903/files#diff-0eb72bc636980e431e5fe14b2a54616dR51

File: src/js/structs/Item.js (hunk 4 above)
Comment (Poltergeist) [inline src/js/structs/Item.js]: Do you mean `this.get('_itemData')` The whole point is to get the whole object which means also unknown values.

Comment (Poltergeist) [inline src/js/structs/Item.js]: Forget my comment, I just realized that `this.get()` will give you `_itemData`.

File: src/js/structs/ServiceTree.js (hunk 2 above)
Comment (jfurrow) [inline src/js/structs/ServiceTree.js]: Done!

File: src/js/structs/Item.js (hunk 4 above)
Comment (rcorral) [inline src/js/structs/Item.js]: A simple test would be great!

File: src/js/utils/ServiceUtil.js (hunk 3 above)
Comment (jfurrow) [inline src/js/utils/ServiceUtil.js]: Done!

File: src/js/utils/ServiceUtil.js (hunk 5 above)
Comment (jfurrow) [inline src/js/utils/ServiceUtil.js]: Done!

File: src/js/utils/ServiceUtil.js (hunk 6 above)
Comment (jfurrow) [inline src/js/utils/ServiceUtil.js]: Removed entirely! :)

File: src/js/utils/ServiceUtil.js (hunk 7 above)
Comment (jfurrow) [inline src/js/utils/ServiceUtil.js]: Done!

File: src/js/structs/Item.js (hunk 4 above)
Comment (jfurrow) [inline src/js/structs/Item.js]: All of these are done :)

File: src/js/utils/ServiceUtil.js (hunk 8 above)
Comment (jfurrow) [inline src/js/utils/ServiceUtil.js]: Done!

File: src/js/constants/BlackListServiceConfig.js (hunk 9 above)
Comment (jfurrow) [inline src/js/constants/BlackListServiceConfig.js]: Done!

File: src/js/utils/ServiceUtil.js (hunk 10 above)
Comment (jfurrow) [inline src/js/utils/ServiceUtil.js]: Done!

File: src/js/components/modals/ServiceFormModal.js (hunk 11 above)
Comment (jfurrow) [inline src/js/components/modals/ServiceFormModal.js]: Done!

File: src/js/constants/blacklistServiceConfig.js (hunk 1 above)
Comment (jfurrow) [inline src/js/constants/blacklistServiceConfig.js]: Done!

File: src/js/constants/BlackListServiceConfig.js (hunk 12 above)
Comment (jfurrow) [inline src/js/constants/BlackListServiceConfig.js]: Yes, it's necessary. Without it we'll be passing the `_itemData` to the JSON editor:

**Comment by jfurrow** [comment]
@Poltergeist @orlandohohmeier @mlunoe I believe I've addressed all of the comments, but let me know if I've missed anything.

File: src/js/structs/ServiceTree.js [hunk 13]
```diff
@@ -34,7 +34,9 @@ module.exports = class ServiceTree extends Tree {
     // Converts items into instances of ServiceTree, Application or Framework
     // based on their properties.
     this.list = this.list.map((item) => {
-      if (item instanceof ServiceTree) {
+      if (item instanceof ServiceTree
+        || item instanceof Framework
```
Comment (mlunoe) [inline src/js/structs/ServiceTree.js]: Check for `Service`

File: src/js/components/modals/ServiceFormModal.js [hunk 14]
```diff
@@ -103,6 +104,15 @@ const responseAttributePathToFieldIdMap = {
   '/': 'general'
 };
 
+var cleanJSONdefinition = function (jsonDefinition) {
+  return Object.keys(jsonDefinition).filter(function (key) {
+    return !ServiceConfig.BLACK_LIST.includes(key);
+  }).reduce(function (memo, key) {
+    memo[key] = jsonDefinition[key];
+    return memo;
```
Comment (mlunoe) [inline src/js/components/modals/ServiceFormModal.js]: Newline before `return`

File: src/js/constants/ServiceConfig.js [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+    'version',
+    'versionInfo',
+    'deployments',
+    'tasks',
+    'taskStats',
+    'tasksHealthy',
+    'tasksRunning',
+    'tasksStaged',
+    'tasksUnhealthy',
+    '_itemData'
```
Comment (mlunoe) [inline src/js/constants/ServiceConfig.js]: This should still not be the case. This means that we are instantiating a struct with struct data, and not raw data

File: src/js/utils/ServiceUtil.js [hunk 16]
```diff
@@ -194,7 +195,7 @@ const ServiceUtil = {
 
           if (localVolumes.length > 0) {
             volumesList = volumesList.concat(localVolumes);
-            if (!isEdit) {
+            if (!isEdit && volumesList.length > 0) {
```
Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: `&& volumesList.length` is enough here

Comment (mlunoe) [inline src/js/utils/ServiceUtil.js]: Done

File: src/js/components/modals/ServiceFormModal.js (hunk 14 above)
Comment (mlunoe) [inline src/js/components/modals/ServiceFormModal.js]: done

File: src/js/structs/ServiceTree.js (hunk 13 above)
Comment (mlunoe) [inline src/js/structs/ServiceTree.js]: done

**Comment by mlunoe** [comment]
I didn't get time to finish this up completely today. This: https://github.com/dcos/dcos-ui/pull/903#discussion_r73409725 is still a big issue, and should be fixed before this PR lands. This https://github.com/dcos/dcos-ui/pull/903/commits/1c092c18c3156cbe1412253be55213d77ab7f4a2 might have fixed it, but I did not verify. Cheers

**Comment by Poltergeist** [comment]
Yeah Looks good can we merge this?

File: src/js/constants/ServiceConfig.js [hunk 17]
```diff
@@ -0,0 +1,17 @@
+let ServiceConfig = {
+  BLACK_LIST: [
```
Comment (pierlo-upitup) [inline src/js/constants/ServiceConfig.js]: Take it or leave it, but it's `blacklist` in one word ;)

**Comment by pierlo-upitup** [comment]
When I create a new service, and switch to JSON immediately, I see 

image

File: src/js/stores/DCOSStore.js [hunk 18]
```diff
@@ -350,6 +354,12 @@ class DCOSStore extends EventEmitter {
         options = Object.assign(options, frameworks[item.getName()]);
       }
 
+      if (item instanceof Item) {
+        return new item.constructor(
```
Comment (pierlo-upitup) [inline src/js/stores/DCOSStore.js]: Why isn't `new Item(...` enough here?

Comment (pierlo-upitup) [inline src/js/stores/DCOSStore.js]: oh I see.

**Comment by Poltergeist** [comment]
This is a bug on master. @pierlo-upitup

**Comment by Poltergeist** [comment]
The bug was possibly introduced with https://github.com/dcos/dcos-ui/pull/889 Maybe @wavesoft has a quick solution to this bug. I do not think that this should block the merge since it is already on master.

**Comment by pierlo-upitup** [comment]
Adding `hold_merge` so @Poltergeist can push a quick fix for https://github.com/dcos/dcos-ui/pull/903#issuecomment-237580553

Besides that, LGTM

**Comment by mlunoe** [comment]
I get this error, if I delete id:

Also this, if I try to switch back to the form:

**Comment by mlunoe** [comment]
This bug persists:

File: src/js/components/modals/ServiceFormModal.js [hunk 19]
```diff
@@ -258,13 +270,15 @@ class ServiceFormModal extends mixin(StoreMixin) {
     }
 
     if (this.state.jsonMode) {
-      let jsonDefinition = this.state.jsonDefinition;
-      marathonAction(JSON.parse(jsonDefinition), this.state.force);
+      let jsonDefinition = JSON.parse(this.state.jsonDefinition);
+      jsonDefinition = cleanJSONdefinition(jsonDefinition);
+      marathonAction(jsonDefinition, this.state.force);
+      jsonDefinition = JSON.stringify(jsonDefinition, null, 2);
```
Comment (pierlo-upitup) [inline src/js/components/modals/ServiceFormModal.js]: Isn't this a bit convoluted?
1) parse the json string
2) remove fields from it and convert to object
3) (call the action)
4) re-convert parsed json back to string
...

File: src/js/structs/Item.js [hunk 20]
```diff
@@ -14,4 +14,8 @@ module.exports = class Item {
 
     return this._itemData[key];
   }
+
+  toJSON() {
```
Comment (orlandohohmeier) [inline src/js/structs/Item.js]: Let's move this to `Service`, as we're introducing it exclusively for the services.

Comment (pierlo-upitup) [inline src/js/structs/Item.js]: Won't this be helpful for Jobs too?

Comment (orlandohohmeier) [inline src/js/structs/Item.js]: Yeah, but I'd rather replicate it or introduce a new shared parent class than adding the method to all(!) of our struct.

File: src/js/components/modals/ServiceFormModal.js (hunk 19 above)
Comment (orlandohohmeier) [inline src/js/components/modals/ServiceFormModal.js]: We're probably able to drastically simplify the whole `handleSubmit` process. I wonder if we should do this in a separate PR. I'd love if we could simplify it in the same way we've did it for the jobs (https://github.com/dcos/dcos-ui/blob/master/src/js/components/modals/JobFormModal.js#L157).

**Comment by orlandohohmeier** [comment]
Let's merge this beast once the tests pass
